File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    const size = s.length;
    if(size === 0) return '';

    let start = 0, maxLength = 1;
    const dp = Array.from({length: size}, () => new Array(size).fill(false));
    console.log(dp);
    for(let i = 0; i < size; i++){
        dp[i][i] = true;
    }

    for(let i = 2; i <= size; i++){
        for(let left = 0; left <= size - i; left++){
            const right = i + left - 1;
            if(s[left] === s[right]){
               if(i === 2) dp[left][right] = true;
               else dp[left][right] = dp[left + 1][right - 1];

               if(dp[left][right] && i > maxLength){
                start = left;
                maxLength = i;
               }
            }
        }
    }
    return s.substring(start, start + maxLength);
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int size = s.size();
        if(size == 0) return "";
        int start(0), maxLength(1);
        vector<vector<bool>> dp(size, vector<bool>(size, false));
        for(int i = 0; i < size; i++){
            dp[i][i] = true;
        }

        for(int len = 2; len <= size; len++){
            for(int left = 0; left <= size - len; left++){
                int right = len + left - 1;
                if(s[left] == s[right]){
                    if(len == 2) dp[left][right] = true;
                    else dp[left][right] = dp[left + 1][right - 1];

                    if(dp[left][right] && len > maxLength){
                        start = left;
                        maxLength = len;
                    }
                }
            }
        }
        return s.substr(start, maxLength);
    }
};